File: sdk/ts/src/blockchain/evm/locking_client.ts
/**
 * Locking Contract Client
 * Provides lock/relock/unlock/withdraw functionality for the NonSlashableAppRegistry contract
 */

import { Address } from 'viem';
import Decimal from 'decimal.js';
import { decimalToBigInt } from '../../core/utils';
import { EVMClient, WalletSigner } from './interface';
import { AppRegistryAbi } from './app_registry_abi';
import { Erc20Abi } from './erc20_abi';

/**
 * LockingClient provides methods to interact with the Locking (NonSlashableAppRegistry) contract.
 * Supports lock, relock, unlock, withdraw, and approve operations.
 */
export class LockingClient {
  private contractAddress: Address;
  private evmClient: EVMClient;
  private walletSigner?: WalletSigner;

  private tokenAddress?: Address;
  private tokenDecimals?: number;

  constructor(contractAddress: Address, evmClient: EVMClient, walletSigner?: WalletSigner) {
    this.contractAddress = contractAddress;
    this.evmClient = evmClient;
    this.walletSigner = walletSigner;
  }

  private requireWalletSigner(): WalletSigner {
    if (!this.walletSigner) {
      throw new Error('Write operations require a wallet signer. In Node.js, use a TransactionSigner that implements getAccount() (e.g., EthereumRawSigner)');
    }
    return this.walletSigner;
  }

  /**
   * Lazily fetch and cache the token address and decimals from the contract.
   */
  private async ensureTokenInfo(): Promise<{ address: Address; decimals: number }> {
    if (this.tokenAddress !== undefined && this.tokenDecimals !== undefined) {
      return { address: this.tokenAddress, decimals: this.tokenDecimals };
    }

    const tokenAddress = (await this.evmClient.readContract({
      address: this.contractAddress,
      abi: AppRegistryAbi,
      functionName: 'asset',
    })) as Address;

    const decimals = (await this.evmClient.readContract({
      address: tokenAddress,
      abi: Erc20Abi,
      functionName: 'decimals',
    })) as number;

    this.tokenAddress = tokenAddress;
    this.tokenDecimals = decimals;
    return { address: tokenAddress, decimals };
  }

  /**
   * Lock tokens into the Locking contract for the specified target address.
   * The caller must have approved the contract to spend tokens beforehand.
   *
   * @param target - The address to lock tokens for
   * @param amount - The amount of tokens to lock (human-readable decimals)
   * @returns Transaction hash
   */
  async lock(target: Address, amount: Decimal): Promise<string> {
    const walletSigner = this.requireWalletSigner();
    const { decimals } = await this.ensureTokenInfo();
    const amountBig = decimalToBigInt(amount, decimals);

    const { request } = await this.evmClient.simulateContract({
      address: this.contractAddress,
      abi: AppRegistryAbi,
      functionName: 'lock',
      args: [target, amountBig],
      account: walletSigner.account!.address,
    });

    const hash = await walletSigner.writeContract(request as any);
    await this.evmClient.waitForTransactionReceipt({ hash });
    return hash;
  }

  /**
   * Re-lock tokens that are in the unlocking state back to the locked state.
   *
   * @returns Transaction hash
   */
  async relock(): Promise<string> {
    const walletSigner = this.requireWalletSigner();
    const { request } = await this.evmClient.simulateContract({
      address: this.contractAddress,
      abi: AppRegistryAbi,
      functionName: 'relock',
      account: walletSigner.account!.address,
    });

    const hash = await walletSigner.writeContract(request as any);
    await this.evmClient.waitForTransactionReceipt({ hash });
    return hash;
  }

  /**
   * Initiate the unlock process for the caller's locked tokens.
   * After the unlock period elapses, withdraw can be called.
   *
   * @returns Transaction hash
   */
  async unlock(): Promise<string> {
    const walletSigner = this.requireWalletSigner();
    const { request } = await this.evmClient.simulateContract({
      address: this.contractAddress,
      abi: AppRegistryAbi,
      functionName: 'unlock',
      account: walletSigner.account!.address,
    });

    const hash = await walletSigner.writeContract(request as any);
    await this.evmClient.waitForTransactionReceipt({ hash });
    return hash;
  }

  /**
   * Withdraw unlocked tokens to the specified destination address.
   * Can only be called after the unlock period has elapsed.
   *
   * @param destination - The address to receive the withdrawn tokens
   * @returns Transaction hash
   */
  async withdraw(destination: Address): Promise<string> {
    const walletSigner = this.requireWalletSigner();
    const { request } = await this.evmClient.simulateContract({
      address: this.contractAddress,
      abi: AppRegistryAbi,
      functionName: 'withdraw',
      args: [destination],
      account: walletSigner.account!.address,
    });

    const hash = await walletSigner.writeContract(request as any);
    await this.evmClient.waitForTransactionReceipt({ hash });
    return hash;
  }

  /**
   * Approve the Locking contract to spend the specified amount of tokens.
   * This must be called before lock.
   *
   * @param amount - The amount of tokens to approve (human-readable decimals)
   * @returns Transaction hash
   */
  async approveToken(amount: Decimal): Promise<string> {
    const walletSigner = this.requireWalletSigner();
    const { address: tokenAddress, decimals } = await this.ensureTokenInfo();
    const amountBig = decimalToBigInt(amount, decimals);

    const { request } = await this.evmClient.simulateContract({
      address: tokenAddress,
      abi: Erc20Abi,
      functionName: 'approve',
      args: [this.contractAddress, amountBig],
      account: walletSigner.account!.address,
    });

    const hash = await walletSigner.writeContract(request as any);
    await this.evmClient.waitForTransactionReceipt({ hash });
    return hash;
  }

  /**
   * Get the locked balance of a user in the Locking contract.
   *
   * @param user - The address to check
   * @returns The locked balance as a Decimal (adjusted for token decimals)
   */
  async getBalance(user: Address): Promise<Decimal> {
    const { decimals } = await this.ensureTokenInfo();

    const balance = (await this.evmClient.readContract({
      address: this.contractAddress,
      abi: AppRegistryAbi,
      functionName: 'balanceOf',
      args: [user],
    })) as bigint;

    return new Decimal(balance.toString()).div(Decimal.pow(10, decimals));
  }

  /**
   * Get the lock state of a user.
   * Returns 0 for None, 1 for Locked, 2 for Unlocking.
   *
   * @param user - The address to check
   * @returns Lock state (0=None, 1=Locked, 2=Unlocking)
   */
  async getLockState(user: Address): Promise<number> {
    return (await this.evmClient.readContract({
      address: this.contractAddress,
      abi: AppRegistryAbi,
      functionName: 'lockStateOf',
      args: [user],
    })) as number;
  }

  /**
   * Get the token decimals used by the Locking contract's asset.
   *
   * @returns Number of decimals
   */
  async getTokenDecimals(): Promise<number> {
    const { decimals } = await this.ensureTokenInfo();
    return decimals;
  }
}

File: pkg/blockchain/evm/locking_client.go
package evm

import (
	"math/big"

	"github.com/ethereum/go-ethereum/accounts/abi/bind"
	"github.com/ethereum/go-ethereum/common"
	"github.com/pkg/errors"
	"github.com/shopspring/decimal"

	"github.com/layer-3/nitrolite/pkg/core"
	"github.com/layer-3/nitrolite/pkg/sign"
)

// LockingClient provides access to a Locking contract.
type LockingClient struct {
	BaseClient
	lockingContractAddress common.Address
	transactOpts           *bind.TransactOpts

	tokenAddress  common.Address
	tokenDecimals uint8
}

// NewLockingClient creates a new LockingClient.
// If txSigner is provided, the client can perform write operations (lock, relock, unlock, withdraw).
func NewLockingClient(lockingContractAddress common.Address, evmClient EVMClient, blockchainID uint64, txSigner ...sign.Signer) (*LockingClient, error) {
	c := &LockingClient{
		BaseClient: BaseClient{
			evmClient:    evmClient,
			blockchainID: blockchainID,
		},
		lockingContractAddress: lockingContractAddress,
	}
	if len(txSigner) > 0 && txSigner[0] != nil {
		c.transactOpts = signerTxOpts(txSigner[0], blockchainID)
	}

	lockingContract, err := NewAppRegistry(c.lockingContractAddress, c.evmClient)
	if err != nil {
		return nil, errors.Wrap(err, "failed to instantiate Locking contract")
	}

	tokenAddress, err := lockingContract.Asset(nil)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get asset from the Locking contract")
	}

	erc20Contract, err := NewIERC20(tokenAddress, c.evmClient)
	if err != nil {
		return nil, errors.Wrap(err, "failed to instantiate ERC20 contract")
	}

	decimals, err := erc20Contract.Decimals(nil)
	if err != nil {
		return nil, errors.Wrapf(err, "failed to get decimals for token %s", tokenAddress.Hex())
	}

	c.tokenAddress = tokenAddress
	c.tokenDecimals = decimals

	return c, nil
}

// GetTokenDecimals returns the number of decimals for the token used in the AppRegistry.
// This is needed to convert between human-readable amounts and the raw integer amounts used in transactions.
func (c *LockingClient) GetTokenDecimals() (uint8, error) {
	lockingContract, err := NewAppRegistry(c.lockingContractAddress, c.evmClient)
	if err != nil {
		return 0, errors.Wrap(err, "failed to instantiate Locking contract")
	}

	tokenAddress, err := lockingContract.Asset(&bind.CallOpts{})
	if err != nil {
		return 0, errors.Wrap(err, "failed to get asset from the Locking contract")
	}

	erc20Contract, err := NewIERC20(tokenAddress, c.evmClient)
	if err != nil {
		return 0, errors.Wrap(err, "failed to instantiate ERC20 contract")
	}

	decimals, err := erc20Contract.Decimals(&bind.CallOpts{})
	if err != nil {
		return 0, errors.Wrapf(err, "failed to get decimals for token %s", tokenAddress.Hex())
	}

	return decimals, nil
}

// Lock locks tokens into the Locking contract for the specified target address.
// The caller must have approved the Locking contract to spend the token beforehand.
func (c *LockingClient) Lock(targetWalletAddress string, amount decimal.Decimal) (string, error) {
	if !common.IsHexAddress(targetWalletAddress) {
		return "", errors.Errorf("invalid address %q", targetWalletAddress)
	}

	targetAddr := common.HexToAddress(targetWalletAddress)
	if c.transactOpts == nil {
		return "", errors.New("transaction signer not configured")
	}

	amountBig, err := core.DecimalToUint256(amount, c.tokenDecimals)
	if err != nil {
		return "", errors.Wrap(err, "failed to convert amount with decimal precision")
	}

	lockingContract, err := NewAppRegistry(c.lockingContractAddress, c.evmClient)
	if err != nil {
		return "", errors.Wrap(err, "failed to instantiate Locking contract")
	}

	tx, err := lockingContract.Lock(c.transactOpts, targetAddr, amountBig)
	if err != nil {
		return "", errors.Wrap(err, "failed to send lock transaction")
	}
	return tx.Hash().Hex(), nil
}

// Relock re-locks tokens that are in the unlocking state back to the locked state.
func (c *LockingClient) Relock() (string, error) {
	if c.transactOpts == nil {
		return "", errors.New("transaction signer not configured")
	}

	lockingContract, err := NewAppRegistry(c.lockingContractAddress, c.evmClient)
	if err != nil {
		return "", errors.Wrap(err, "failed to instantiate Locking contract")
	}

	tx, err := lockingContract.Relock(c.transactOpts)
	if err != nil {
		return "", errors.Wrap(err, "failed to send relock transaction")
	}
	return tx.Hash().Hex(), nil
}

// Unlock initiates the unlock process for the caller's locked tokens.
// After the unlock period elapses, Withdraw can be called.
func (c *LockingClient) Unlock() (string, error) {
	if c.transactOpts == nil {
		return "", errors.New("transaction signer not configured")
	}

	lockingContract, err := NewAppRegistry(c.lockingContractAddress, c.evmClient)
	if err != nil {
		return "", errors.Wrap(err, "failed to instantiate Locking contract")
	}

	tx, err := lockingContract.Unlock(c.transactOpts)
	if err != nil {
		return "", errors.Wrap(err, "failed to send unlock transaction")
	}
	return tx.Hash().Hex(), nil
}

// Withdraw withdraws unlocked tokens to the specified destination address.
// Can only be called after the unlock period has elapsed.
func (c *LockingClient) Withdraw(destination string) (string, error) {
	if c.transactOpts == nil {
		return "", errors.New("transaction signer not configured")
	}

	destinationAddr := common.HexToAddress(destination)

	lockingContract, err := NewAppRegistry(c.lockingContractAddress, c.evmClient)
	if err != nil {
		return "", errors.Wrap(err, "failed to instantiate Locking contract")
	}

	tx, err := lockingContract.Withdraw(c.transactOpts, destinationAddr)
	if err != nil {
		return "", errors.Wrap(err, "failed to send withdraw transaction")
	}
	return tx.Hash().Hex(), nil
}

// ApproveToken approves the Locking contract to spend the specified amount of tokens.
// This must be called before Lock.
func (c *LockingClient) ApproveToken(amount decimal.Decimal) (string, error) {
	if c.transactOpts == nil {
		return "", errors.New("transaction signer not configured")
	}

	amountBig, err := core.DecimalToUint256(amount, c.tokenDecimals)
	if err != nil {
		return "", errors.Wrap(err, "failed to convert amount with decimal precision")
	}

	lockingContract, err := NewAppRegistry(c.lockingContractAddress, c.evmClient)
	if err != nil {
		return "", errors.Wrap(err, "failed to instantiate Locking contract")
	}

	tokenAddress, err := lockingContract.Asset(&bind.CallOpts{})
	if err != nil {
		return "", errors.Wrap(err, "failed to get asset from Locking contract")
	}

	erc20Contract, err := NewIERC20(tokenAddress, c.evmClient)
	if err != nil {
		return "", errors.Wrap(err, "failed to instantiate ERC20 contract")
	}

	tx, err := erc20Contract.Approve(c.transactOpts, c.lockingContractAddress, amountBig)
	if err != nil {
		return "", errors.Wrap(err, "failed to send approve transaction")
	}
	return tx.Hash().Hex(), nil
}

// GetBalance returns the locked balance of a user in the Locking contraсt.
func (c *LockingClient) GetBalance(user string) (decimal.Decimal, error) {
	userAddr := common.HexToAddress(user)
	lockingContract, err := NewAppRegistry(c.lockingContractAddress, c.evmClient)
	if err != nil {
		return decimal.Zero, errors.Wrap(err, "failed to instantiate Locking contract")
	}

	balance, err := lockingContract.BalanceOf(nil, userAddr)
	if err != nil {
		return decimal.Zero, errors.Wrap(err, "failed to get balance")
	}

	return decimal.NewFromBigInt(balance, -int32(c.tokenDecimals)), nil
}

// decimalToBigInt converts a decimal amount to *big.Int given token decimals.
func decimalToBigInt(amount decimal.Decimal, decimals int32) *big.Int {
	shifted := amount.Shift(decimals)
	return shifted.BigInt()
}
